00_Base/src/datasources/TokensDatasource.ts

```diff
@@ -0,0 +1,229 @@
+import { TokensRepository } from '../repository/TokensRepository';
+import { IdTokenEnumType } from '@citrineos/base';
+import { UnknownTokenException } from '../exception/unknown.token.exception';
+import { TokensValidators } from '../util/validators/TokensValidators';
+import { Service } from 'typedi';
+
+@Service()
+export class TokensDatasource implements ITokensDatasource {
+  constructor(
+    private readonly authorizationRepository: SequelizeAuthorizationRepository,
+    private readonly tokensRepository: TokensRepository,
+    private readonly logger: OcpiLogger,
+  ) {}
+
+  async getSingleToken(
+    tokenRequest: SingleTokenRequest,
+  ): Promise<TokenDTO | undefined> {
+    try {
+      const ocppAuths = await this.authorizationRepository.readAllByQuerystring(
+        {
+          idToken: tokenRequest.uid,
+          type: OcpiTokensMapper.mapOcpiTokenTypeToOcppIdTokenType(
+            tokenRequest?.type ?? TokenType.RFID,
+          ),
+        },
+      );
+
+      if (ocppAuths.length === 0) {
+        return undefined;
+      }
+
+      const ocpiToken = await this.tokensRepository.readOnlyOneByQuery({
+        where: {
+          authorization_id: {
+            [Op.in]: ocppAuths.map((ocppAuth) => ocppAuth.id),
+          },
+          country_code: tokenRequest.country_code,
+          party_id: tokenRequest.party_id,
+          type: tokenRequest.type,
+        },
+      });
+
+      if (!ocpiToken) {
+        return undefined;
+      }
+
+      return OcpiTokensMapper.toDto(
+        this.getMatchingAuth(ocpiToken.authorization_id, ocppAuths)!,
+        ocpiToken,
+      );
+    } catch (error) {
+      this.logger.error('Error retrieving single token', error);
+      throw error;
+    }
+  }
+
+  async patchToken(partialToken: TokenDTO): Promise<TokenDTO> {
+    TokensValidators.validatePartialTokenForUniquenessRequiredFields(
+      partialToken,
+    );
+
+    const ocppAuths = await this.authorizationRepository.readAllByQuerystring({
+      idToken: partialToken.uid!,
+      type: OcpiTokensMapper.mapOcpiTokenTypeToOcppIdTokenType(
+        partialToken.type,
+      ),
+    });
+
+    if (ocppAuths.length === 0) {
+      throw new UnknownTokenException('Token not found in the database');
+    }
+
+    const existingOcpiToken = await this.tokensRepository.readOnlyOneByQuery({
+      where: {
+        authorization_id: {
+          [Op.in]: ocppAuths.map((ocppAuth) => ocppAuth.id),
+        },
+        country_code: partialToken.country_code,
+        party_id: partialToken.party_id,
+        type: partialToken.type,
+      },
+    });
+
+    if (!existingOcpiToken) {
+      throw new UnknownTokenException('Token not found in the database');
+    }
+
+    const updatedToken = await this.tokensRepository.updateByKey(
+      partialToken,
+      existingOcpiToken.id,
+    );
+
+    if (!updatedToken) {
+      throw new UnknownTokenException('Token not found in the database');
+    }
+    //
+    // const newOCPPAuth =
```

Comment:
Either use it or delete it. Commenting out code usually just makes things harder to understand ;)

Reply:
Removed for now